test/test_common/simulated_time_system.cc

```diff
@@ -308,34 +279,40 @@ Thread::CondVar::WaitStatus SimulatedTimeSystemHelper::waitFor(Thread::MutexBasi
   return Thread::CondVar::WaitStatus::Timeout;
 }
 
-MonotonicTime
-SimulatedTimeSystemHelper::alarmTimeLockHeld(Alarm* alarm) ABSL_NO_THREAD_SAFETY_ANALYSIS {
-  // We disable thread-safety analysis as the compiler can't detect that
-  // alarm_->timeSystem() == this, so we must be holding the right mutex.
-  ASSERT(&(alarm->timeSystem()) == this);
-  return alarm->time();
-}
-
 void SimulatedTimeSystemHelper::alarmActivateLockHeld(Alarm* alarm) ABSL_NO_THREAD_SAFETY_ANALYSIS {
   // We disable thread-safety analysis as the compiler can't detect that
   // alarm_->timeSystem() == this, so we must be holding the right mutex.
   ASSERT(&(alarm->timeSystem()) == this);
   alarm->activateLockHeld();
 }
 
-int64_t SimulatedTimeSystemHelper::nextIndex() {
-  absl::MutexLock lock(&mutex_);
-  return index_++;
-}
-
 void SimulatedTimeSystemHelper::addAlarmLockHeld(
     Alarm* alarm, const std::chrono::microseconds& duration) ABSL_NO_THREAD_SAFETY_ANALYSIS {
   ASSERT(&(alarm->timeSystem()) == this);
-  alarm->setTimeLockHeld(monotonic_time_ + duration);
-  alarms_.insert(alarm);
+  ASSERT(alarms_.size() == alarm_registrations_map_.size());
+  ASSERT(alarm_registrations_map_.find(alarm) == alarm_registrations_map_.end());
+
+  auto insert_result = alarms_.insert({monotonic_time_ + duration, random_source_.random(), alarm});
```

Comment:
I found it hard to find this initialization line. WDYT of giving AlarmRegistration a ctor:
```
  AlarmRegistration(MonotonicTime time, uint64 randomness, Alarm* alarm) : ...
```
so that it was more obvious from looking at the class how the member variables must always be initialized.

Reply:
IIRC there's plenty of prior art in Envoy for use of the c++ 11 struct initializer syntax in these cases.

Added constructor implementation.